docs/modules/architecture/pages/compact-binary-specification.adoc

```diff
@@ -0,0 +1,706 @@
+
+== Variable-Size Fields
+
+The offsets of variable-size fields are written at the end in the alphabetical order of the field names. To read a variable-size field from the data, one should read the index of the offset from the Schema. Then read the related index is read from the end of the data to get the offset. The variable-size field can be read from this offset.
+
+On the schema, the index for a variable-size field is determined as follows:
+
+* The fields are given the index incrementally according to the order of the field names starting from 0.
+
+Based on the length of the serialized data, the offsets of the variable-size fields might be represented by 1, 2, or 4 bytes. Note that, this does not mean that offsets will be represented by variable-size integers. It simply means that all variable-size field offsets will either be 1, 2, or 4 bytes per serialized object, depending on its size.
+
+== Schema
+
+[cols="1,1"]
+|===
+|Name |Type
+|type name |string
+|number of fields |i32
+|name of field 0 |string
+|typeid of field 0 |i32
+|name of field 1 |string
+|typeid of field 1 |i32
+|... |...
+|name of field n |string
+|typeid of field n |i32
+|===
+
+When writing a schema to the wire, fields will be ordered according to their name so that the same structure will result in the same byte representation and produce the same schema id.
+
+The offsets and indexes are also decided on the ordered fields. The smaller-sized fields come first. The order is by name within the same size fields.
+
+In the Schema class, each field will either
+
+* have a positive offset, if it is a fixed-size field
+* have a positive index if it is a variable-size field
+* have a positive bit offset if it is a boolean field. This bit offset is the offset within the byte given by the normal offset
+
+== Schema ID
+
+We are using 64bit https://en.wikipedia.org/wiki/Rabin_fingerprint[Rabin fingerprint] to create [.inline-comment-marker]#a schema id#.
+
+Rabin fingerprint is chosen mostly because it is recommended in Avro’s documents as follows +
+https://avro.apache.org/docs/current/spec.html#schema_primitive
+
+[source,text]
+----
+At the opposite extreme, the smallest fingerprint recommended is a 64-bit Rabin fingerprint. Below, there is a provided pseudo-code for this algorithm that can be easily translated into any programming language. 64-bit fingerprints should guarantee uniqueness for schema caches of up to a million entries (for such a cache, the chance of a collision is 3E-8). It is not recommended to use shorter fingerprints, as the chances of collisions are too high (for example, with 32-bit fingerprints, a cache with as few as 100,000 schemas has a 50% chance of having a collision).
```

Comment:
What does "opposite extreme" mean here in this context?

Reply:
The paragraph is from https://avro.apache.org/docs/1.11.1/specification/_print/ . It means high number of bits fingerprint algorithms as opposed to low number of bits. (e.g 256bit instead of 64)


What do you think we should change here?